# Python for Data Science Practicce Session 1: Social Sciences

## How is Coronavirus Affecting Lives Across the World?

The [COVIDiSTRESS global survey](https://osf.io/z39us/) is an international collaborative undertaking for gathering data on human experiences, behaviour and attitudes during the COVID-19 pandemic. The survey focuses on psychological stress, compliance with behavioural guidelines to slow the spread of Coronavirus, and trust in governmental institutions and their preventive measures. 

In this notebook we are going to use the results of the COVIDiSTRESS survey conducted in April–May 2020 to investigate the impact of the Coronavirus pandemic on the perceived level of stress by individuals across the world as well as explore other related factors. 

Before we get started we need to import the `pandas` packages

In [152]:
# Import pandas
import pandas as pd

## Data Preprocessing

### Import

You can download the relevant dataset as a CSV file from this session's [materials](https://education.wdss.io/python-for-data-science/session-one/).

This dataset is more complex than the one we looked at in the teaching session and so we will need to acquaint ourselves with some more optional arguments of the `read_csv` function. Namely, look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) of the function, to see what the following parameters are for:

- `low_memory`
- `parse_dates`
- `dayfirst`
- `index_col`

Use this knowledge to import the dataset and assign it to the variable `raw_df`. Pay close attention to the following points:

- The dataset has a date column `RecordedDate` which uses day first format
- Because many columns have lots of missing values, there will likely be mixed type inference (a bad thing) when using low memory mode
- The first column of the dataset consists of row numbers which should be read as an index

In [154]:
raw_df = pd.read_csv(
    'covidstress.csv',
    index_col=0,
    dayfirst=True,
    low_memory=False,
    parse_dates=['RecordedDate']
)

> **Bonus**
>
> Suppose we didn't trust the index column of the original dataset. How could we use the default index provide by `pandas` instead?

Let's explore the dataset by doing the following:
- Print the first 3 rows of the dataset
- Print the dimensions of the dataset
- Print numerical summaries of the columns
- Print the data types of the columns

In [5]:
# First 3 rows
raw_df.head(3)

Duration..in.seconds.        RecordedDate UserLanguage  Dem_age Dem_gender  \
1                    180 2020-05-30 23:47:00          SAR       29     Female   
2                   3100 2020-05-29 23:30:00           UR       20       Male   
3                    127 2020-05-30 22:40:00          SAR       47     Female   

                                             Dem_edu  \
1                   College degree, bachelor, master   
2                   College degree, bachelor, master   
3  Some College, short continuing education or eq...   

                  Dem_edu_mom Dem_employment    Country Dem_Expat  ...  \
1  Some College or equivalent   Not employed  Argentina       yes  ...   
2                        None        Student   Pakistan       yes  ...   
3  Some College or equivalent  Self-employed  Argentina        no  ...   

  Final_open PSS10_avg  SLON3_avg  neu  ext       ope  agr  con SPS_avg  \
1        NaN       2.9   3.000000  NaN  NaN       NaN  NaN  NaN     NaN   
2        NaN       2.2   2.333333  2.0  5.0  5.333333  5.0  5.0     5.0   
3        NaN       NaN        NaN  NaN  NaN       NaN  NaN  NaN     NaN   

  Scale_UCLA_TRI_avg  
1                NaN  
2                NaN  
3                NaN  

[3 rows x 152 columns]

In [6]:
# Dimensions of the dataset
raw_df.shape

(65535, 152)

In [7]:
# Numerical summaries of the columns
raw_df.describe()

Duration..in.seconds.       Dem_age  Dem_dependents  \
count           6.553500e+04  65535.000000    63317.000000   
mean            2.125383e+03     40.741375        0.933146   
std             9.888777e+03     14.322085        1.805861   
min             2.900000e+01     18.000000        0.000000   
25%             7.120000e+02     29.000000        0.000000   
50%             1.013000e+03     40.000000        0.000000   
75%             1.454000e+03     51.000000        2.000000   
max             1.040679e+06    110.000000      110.000000   

       Dem_isolation_adults  Dem_isolation_kids  Scale_PSS10_UCLA_1  \
count          53379.000000        52463.000000        60336.000000   
mean               1.533937            0.484513            2.556782   
std                3.231931            1.275804            1.094821   
min                0.000000            0.000000            1.000000   
25%                1.000000            0.000000            2.000000   
50%                1.000000            0.000000            3.000000   
75%                2.000000            1.000000            3.000000   
max              110.000000          110.000000            5.000000   

       Scale_PSS10_UCLA_2  Scale_PSS10_UCLA_3  Scale_PSS10_UCLA_4  \
count        60312.000000        60344.000000        60274.000000   
mean             2.663450            3.078019            3.683612   
std              1.104996            1.112253            1.049348   
min              1.000000            1.000000            1.000000   
25%              2.000000            2.000000            3.000000   
50%              3.000000            3.000000            4.000000   
75%              3.000000            4.000000            4.000000   
max              5.000000            5.000000            5.000000   

       Scale_PSS10_UCLA_5  ...  Expl_media_6     PSS10_avg     SLON3_avg  \
count        60253.000000  ...   48753.00000  60811.000000  60673.000000   
mean             3.158266  ...       3.31034      2.628514      2.581236   
std              0.968456  ...       1.37873      0.735439      0.990043   
min              1.000000  ...       1.00000      1.000000      1.000000   
25%              3.000000  ...       2.00000      2.100000      2.000000   
50%              3.000000  ...       3.00000      2.600000      2.666667   
75%              4.000000  ...       4.00000      3.100000      3.333333   
max              5.000000  ...       6.00000      5.000000      5.000000   

                neu           ext           ope           agr           con  \
count  56884.000000  56858.000000  56856.000000  56869.000000  56865.000000   
mean       3.316275      3.999156      4.478871      4.439443      4.380509   
std        1.050524      1.116921      0.947898      0.832629      0.890780   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        2.666667      3.333333      4.000000      4.000000      3.666667   
50%        3.333333      4.000000      4.666667      4.666667      4.333333   
75%        4.000000      5.000000      5.000000      5.000000      5.000000   
max        6.000000      6.000000      6.000000      6.000000      6.000000   

            SPS_avg  Scale_UCLA_TRI_avg  
count  49669.000000           51.000000  
mean       4.849583            1.818627  
std        0.907215            0.968060  
min        1.000000            0.000000  
25%        4.400000            1.000000  
50%        5.000000            1.750000  
75%        5.500000            2.500000  
max        6.000000            4.000000  

[8 rows x 127 columns]

In [8]:
# Data types of columns
with pd.option_context("display.max_rows", 152, "display.max_columns", 5):
    print(raw_df.dtypes)

Duration..in.seconds.             int64
RecordedDate             datetime64[ns]
UserLanguage                     object
Dem_age                           int64
Dem_gender                       object
Dem_edu                          object
Dem_edu_mom                      object
Dem_employment                   object
Country                          object
Dem_Expat                        object
Dem_state                        object
Dem_maritalstatus                object
Dem_dependents                  float64
Dem_riskgroup                    object
Dem_islolation                   object
Dem_isolation_adults            float64
Dem_isolation_kids              float64
AD_gain                          object
AD_loss                          object
AD_check                         object
Scale_PSS10_UCLA_1              float64
Scale_PSS10_UCLA_2              float64
Scale_PSS10_UCLA_3              float64
Scale_PSS10_UCLA_4              float64
Scale_PSS10_UCLA_5              float64


> **Bonus**
>
> When printing the above, we only get to see 10 columns in total. For printing out full datasets and series, checkout the [option context](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.option_context.html)

### Cleaning

There are 152 columns in the dataset, many of which we won't need in our analysis. Instead, we want to focus on the following variables.

| description | variable name | measurement |
| --- | --- | --- |
| date of record | `RecordedDate` | date "DD/MM/YYYY hh:mm"
| age | `Dem_age` | age in years
| gender | `Dem_gender` | Male/Female
| country | `Country` | country name
| employment status| `Dem_employment` | type of employment
| perceived stress for the past week | `Scale_PSS10_UCLA_1` to `Scale_PSS10_UCLA_10` | 1=never, 5=very often; 10 items
| trust in institutions | `OECD_insititutions_1` (government), `OECD_insititutions_4` (health system) | 0=not at all, 10=completely
| trust in country's preventive measures | `Trust_countrymeasure` | 0=too little, 5= appropriate, 11=too much

Stress was meassured in 10 different categories. Create a list, `stress_columns` of column names `Scale_PSS10_UCLA_X`, where `X` ranges from 1 to 10.

In [9]:
stress_columns = ["Scale_PSS10_UCLA_" + str(i) for i in range(1,11)]

We can then combine these with the other column names.

In [10]:
columns = [
    'RecordedDate', 'Dem_age', 'Dem_gender', 'Country', 'Dem_employment',
    'OECD_insititutions_1', 'OECD_insititutions_4', 'Trust_countrymeasure'
] + stress_columns

Now, create a subset of this dataset, `covid_stress`, containing only columns of interest.

In [11]:
covid_stress = raw_df[columns]

The original dataset is fairly large so we may want to remove it from memory now that we no longer need it.

In [12]:
# Clear up memory
del raw_df

Many fields in the dataset have missing/unknown values. We can use `covid_stress.isna()` to return a Boolean dataset where each value is `True` if the original value in that field was missing. Let's investigate where these missing values are.

> Remember, `True`/`False` are converted to `1`/`0` respectively when cast to integers. This means we can use the `sum` and `mean` methods on a Boolean dataset or series to count or find proportions of `True`'s

In [34]:
# How many missing values are there in each column?
covid_stress.isna().sum(axis=0)

RecordedDate               0
Dem_age                    0
Dem_gender               154
Country                  345
Dem_employment           659
OECD_insititutions_1    8074
OECD_insititutions_4    7547
Trust_countrymeasure    7304
Scale_PSS10_UCLA_1      5199
Scale_PSS10_UCLA_2      5223
Scale_PSS10_UCLA_3      5191
Scale_PSS10_UCLA_4      5261
Scale_PSS10_UCLA_5      5282
Scale_PSS10_UCLA_6      5188
Scale_PSS10_UCLA_7      5277
Scale_PSS10_UCLA_8      5298
Scale_PSS10_UCLA_9      5163
Scale_PSS10_UCLA_10     5190
dtype: int64

In [41]:
# What proportion of records (rows) have at least one missing value?
covid_stress_missing = covid_stress.isna().sum(axis=1)!=0
covid_stress_missing.sum() / len(covid_stress)

0.1706111238269627

Given that we have 65535 records in total we can drop all the records with at least one missing value and still have many tens of thousands left. We will learn how to deal with missing values properly in later sessions.

To do this, we use the `dropna` method of the dataframe. Search for "pandas drop missing values" on the web to find the documentation for this. Save the cleaned dataset as `covid_stress_clean`

In [42]:
covid_stress_clean = covid_stress.dropna()

> **Bonus**
>
> Recall from the teaching session that many pandas operations can be done _inplace_. How would we drop rows with missing values in this fashion (check the documentation again if needed).

Let's take a random sample of 10 rows and verify that they have no missing values.

In [43]:
# Random sample of cleaned dataset
covid_stress_clean.sample(10)

RecordedDate  Dem_age Dem_gender    Country      Dem_employment  \
34022 2020-04-08 09:13:00       28     Female    Finland             Student   
14724 2020-04-17 11:56:00       24     Female  Argentina  Full time employed   
16956 2020-04-17 01:38:00       28     Female    Denmark  Full time employed   
13805 2020-04-18 02:36:00       45     Female    Finland  Full time employed   
15568 2020-04-17 05:50:00       40       Male    Denmark  Full time employed   
54120 2020-04-06 03:02:00       18     Female   Bulgaria       Self-employed   
64999 2020-04-03 16:44:00       40     Female     Mexico       Self-employed   
5971  2020-05-09 11:02:00       32     Female     France        Not employed   
7995  2020-05-05 15:24:00       49     Female     Mexico       Self-employed   
32557 2020-04-08 15:22:00       32       Male    Finland  Full time employed   

       OECD_insititutions_1  OECD_insititutions_4  Trust_countrymeasure  \
34022                   9.0                   8.0                   5.0   
14724                   3.0                   8.0                   4.0   
16956                   8.0                  10.0                   5.0   
13805                   9.0                   9.0                   5.0   
15568                   5.0                   7.0                   5.0   
54120                   3.0                   2.0                   7.0   
64999                   8.0                  10.0                   7.0   
5971                    1.0                   9.0                   1.0   
7995                    1.0                   1.0                   0.0   
32557                   0.0                   5.0                   0.0   

       Scale_PSS10_UCLA_1  Scale_PSS10_UCLA_2  Scale_PSS10_UCLA_3  \
34022                 2.0                 4.0                 3.0   
14724                 5.0                 2.0                 4.0   
16956                 2.0                 2.0                 1.0   
13805                 4.0                 4.0                 5.0   
15568                 4.0                 4.0                 4.0   
54120                 4.0                 5.0                 4.0   
64999                 3.0                 1.0                 3.0   
5971                  4.0                 5.0                 4.0   
7995                  2.0                 1.0                 2.0   
32557                 2.0                 1.0                 2.0   

       Scale_PSS10_UCLA_4  Scale_PSS10_UCLA_5  Scale_PSS10_UCLA_6  \
34022                 3.0                 4.0                 2.0   
14724                 5.0                 3.0                 3.0   
16956                 5.0                 4.0                 2.0   
13805                 2.0                 3.0                 4.0   
15568                 3.0                 2.0                 4.0   
54120                 3.0                 1.0                 2.0   
64999                 4.0                 4.0                 3.0   
5971                  3.0                 3.0                 5.0   
7995                  4.0                 5.0                 2.0   
32557                 5.0                 3.0                 1.0   

       Scale_PSS10_UCLA_7  Scale_PSS10_UCLA_8  Scale_PSS10_UCLA_9  \
34022                 3.0                 3.0                 2.0   
14724                 4.0                 1.0                 5.0   
16956                 5.0                 4.0                 1.0   
13805                 3.0                 2.0                 3.0   
15568                 3.0                 3.0                 3.0   
54120                 2.0                 1.0                 5.0   
64999                 4.0                 4.0                 2.0   
5971                  3.0                 3.0                 5.0   
7995                  4.0                 4.0                 3.0   
32557                 5.0                 5.0                 2.0   

       Scale_PSS10_UCLA_10

> Note, the `covid_stress` dataset isn't too large so we don't need to worry about deleting it. Sometimes it's worth holding onto these things in case we need to go back to them and don't want to rerun cells. How big is the dataset exactly? Have a look at the documentation for the `info` method, which can answer this.

### Feature Engineering

Before we continue, we will make a copy of the current dataframe. This will make it easier to return to this point by rerunning this cell if any mistakes are made.

In [51]:
covid_df = covid_stress_clean.copy()

The levels of stress were measured in 10 different categories on a scale from 1=never, 5=very often. We want analyse only aggregated information about the stress level, i.e. the mean of those 10 values. For this, we create a new column `Avg_stress`

In [52]:
# Create new column by averaging all stress columns
covid_df['Avg_stress']=covid_df[stress_columns].mean(axis=1)

# Drop the now redundant stress columns
covid_df=covid_df.drop(stress_columns,axis=1)

The column names `OECD_insititutions_1`, `OECD_insititutions_4` are not the most friendly. Let's rename them to something more intuitive.

In [53]:
mapper = {
    "OECD_insititutions_1" : "Trust_gov",
    "OECD_insititutions_4" : "Trust_health"
}
covid_df = covid_df.rename(mapper, axis=1)

We're now ready to start answering questions about the dataset. Let's print out the data types of the columns to check that everything is as expected before we continue.

In [54]:
covid_df.dtypes

RecordedDate            datetime64[ns]
Dem_age                          int64
Dem_gender                      object
Country                         object
Dem_employment                  object
Trust_gov                      float64
Trust_health                   float64
Trust_countrymeasure           float64
Avg_stress                     float64
dtype: object

Check that the variable types of each of the column is as expected

## Data Analysis

### High-level Questions

When the study was conducted? (i.e. what is the range of dates in the records?)

In [55]:
# Minimum and maximum dates in dataset
covid_df['RecordedDate'].min()

Timestamp('2020-04-03 14:34:00')

In [56]:
covid_df['RecordedDate'].max()

Timestamp('2020-05-30 20:14:00')

What proportion of observations are from Finland, Sweden, or Norway? Use `.round()` to print the answer to 3 decimal places.

In [72]:
# Proportion of Finish, Swedish, Norwegian respondents
(covid_df['Country'].isin(['Finalnd','Sweden','Norway']).sum()/covid_df.shape[0]).round(3)

0.048

Print the dataset sorted by age in descending order.

In [24]:
# Sort by decreasing age
covid_df.sort_values('Dem_age', ascending=False)

RecordedDate  Dem_age Dem_gender    Country      Dem_employment  \
11825 2020-04-22 21:34:00      110     Female  Argentina       Self-employed   
27914 2020-04-10 12:19:00      108     Female    Finland             Retired   
46264 2020-04-06 03:52:00      101       Male   Colombia             Retired   
10156 2020-04-26 14:11:00       98       Male    Armenia  Part time employed   
11975 2020-04-22 11:29:00       92     Female    Denmark             Retired   
...                   ...      ...        ...        ...                 ...   
9428  2020-04-29 19:12:00       18       Male  Argentina             Student   
44588 2020-04-06 10:12:00       18     Female     France             Student   
28499 2020-04-11 06:45:00       18     Female    Croatia             Student   
63009 2020-04-04 02:15:00       18     Female    Finland             Student   
62164 2020-04-04 04:37:00       18       Male      Spain             Student   

       Trust_gov  Trust_health  Trust_countrymeasure  Avg_stress  
11825        6.0           4.0                   8.0         3.1  
27914        9.0           9.0                   5.0         2.7  
46264        0.0           0.0                   1.0         1.3  
10156        1.0           1.0                  10.0         2.7  
11975        9.0           9.0                   5.0         2.5  
...          ...           ...                   ...         ...  
9428         3.0           7.0                   3.0         3.1  
44588        1.0           4.0                   4.0         2.5  
28499        5.0          10.0                   5.0         2.8  
63009        9.0           9.0                   5.0         3.0  
62164        5.0           8.0                   4.0         2.7  

[54354 rows x 9 columns]

What about the median value for age? Print your answer as part of a meaningful sentence.

In [25]:
# Median age of respondents
print("The median age of respondents is", covid_df.Dem_age.median())

The median age of respondents is 40.0


### Characteristics of Respondents

What is the gender breakdown of respondents? Use the `value_counts` method to find out (after searching for its documentation of course).

In [74]:
# Gender breakdown
covid_df['Dem_gender'].value_counts()

Female                        40286
Male                          13468
Other/would rather not say      600
Name: Dem_gender, dtype: int64

It would be nicer to see percentages rather than absolute counts. There are many ways to do this. Take your pick or write down as many as you can think of.

In [75]:
# Gender breakdown as proportions
covid_df['Dem_gender'].value_counts(normalize=True)

Female                        0.741178
Male                          0.247783
Other/would rather not say    0.011039
Name: Dem_gender, dtype: float64

Suppose we want to divide the respondents in 4 age of equal width, how many respondents would land in each group? (This can also be performed using `value_counts` too)

In [83]:
# Binned value counts
covid_df['Dem_age'].value_counts(bins=4)

(17.907, 41.0]    29455
(41.0, 64.0]      21459
(64.0, 87.0]       3427
(87.0, 110.0]        13
Name: Dem_age, dtype: int64

What proportion of the observations have each employment type, excluding students and retirees?

In [106]:
# Employment types for non-students/retired
covid_df[~covid_df['Dem_employment'].isin(['Student','Retired'])]['Dem_employment'].value_counts(normalize=True)

Full time employed    0.611602
Self-employed         0.138477
Part time employed    0.129149
Not employed          0.120772
Name: Dem_employment, dtype: float64

### Stress During the COVID-19 Pandemic

We can perform grouped summaries using the `groupby` method, documented [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html). These returned a grouped dataframe, which we can apply aggregation methods such as `mean` and `min` to, in which case the aggregation is applied separately to each group. An example is given below.

In [30]:
# First observation for each gender
covid_df.groupby('Dem_gender')['RecordedDate'].min()

Dem_gender
Female                       2020-04-03 14:34:00
Male                         2020-04-03 14:39:00
Other/would rather not say   2020-04-03 15:19:00
Name: RecordedDate, dtype: datetime64[ns]

Identify the 10 countries with the highest average value of stress.

In [149]:
covid_df.groupby('Country')['Avg_stress'].max().sort_values(axis=0,ascending=False).head(10)

Country
Japan            5.0
Finland          4.9
Philippines      4.9
Angola           4.8
Malaysia         4.8
Croatia          4.7
France           4.6
Kosovo           4.5
United States    4.5
Bulgaria         4.5
Name: Avg_stress, dtype: float64

How much can we trust these values? Are there enough respondents from each of those countries to make an an inference about the level of stress of the entire country? It would be helpful to aggregate both the mean stress level and observation count. We can do this using the `agg` method. Read the documentation for this then use it to recreate the above result with counts.

In [ ]:
covid_df.groupby('Country')['Avg_stress']

Suprising? Now let's identify 10 countries with highest average value of stress with at least 200 respondents.

In [31]:
country_df[country_df['Count'] >= 200] \
    .sort_values('Mean_stress', ascending=False).head(10)

NameError: name 'country_df' is not defined

To extend this comparison, we wish to create a dateset `country_fltr` which contains the mean stress, trust in health organisations, government and health measures, and the observation count for all countries with at least 200 observations.

In [ ]:
country_df = covid_df.groupby('Country').agg(
    Mean_stress=('Avg_stress', 'mean'),
    Mean_gov_trust=('Trust_gov', 'mean'),
    Mean_health_trust=('Trust_health', 'mean'),
    Mean_measures_trust=('Trust_countrymeasure', 'mean'),
    Count=('Country', 'count')
)

country_fltr = country_df[country_df.Count>200]

Now sort `country_fltr` by different categories and in ascending/descending order, to explore these 4 measure across the countries.

It might be interesting to compare the trust in country's health system with the average stress level. One way to do it is to make a scatter plot with the mean trust in health on the x-axis and mean level of stress on the y-axis. (More about plotting in next session, but I couldn't resist making at least one plot in this notebook)

In [ ]:
country_fltr.plot(x='Mean_health_trust', y='Mean_stress', kind='scatter');

Cool! It seems like there is a negative correlation between these two variables. Can you spot the outlier? Find out which country this datapoint corresponds to. 

### A Closer Look at the UK

In this section we will draw comparisons between the United Kingdom and the rest of Europe. We have provided an Excel spreadsheet in the [session materials](https://education.wdss.io/python-for-data-science/session-one/) that maps countries to continents, which can be used to aid in this comparison. Start by loading this dataset into the notebook.

In [150]:
# Load the country mapping spreadsheet
countries_by_continent = pd.read_excel('countries_by_continent.xlsx')

ImportError: Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.

Create a pandas series of all European countries except the United Kingdom.

Use this series to create two dataframes, `uk_df` and `euro_df` containing observations from the UK and rest of Europe respectively.

In [ ]:
uk_df = # ...
euro_df = # ...

Is the average stress level in the UK above or below average of the rest of Europe? By how much? 

What proportion of people in the UK aged between 40 and 60 (inclusive) reported an average stress level above 3? This is a longer question so try to break it up into sensible chunks.

What the average stress level for each employment status and gender pair in Europe? (Note: you can use a list of labels when grouping a dataframe)

> Wow! What happened there? We've just been introduced to the pandas multi-index, a relatively unique feature of the package which offers incredible power and flexibility. This is considered an advanced feature and so we will not go into any more detail about it but you can read more [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html).

Did the trust in government tend to increase/decrease or remain stable over the period considered in the report? Compare the mean trust in government in first 3 weeks of April starting from 2020-04-06.

You can do this using base Python `for` loops. Loop through through consecutive pairs of days in the provided date list and calculate the mean trust in government for an subset of the dataset filtered to only have days between the current date in the list and 7 days in the future.

This isn't the most elegant and is certainly not the most efficient. A better solution is below, though it goes to show that pandas and base Python can be used together when needed.

In [ ]:
uk_df.groupby(uk_df.RecordedDate.dt.isocalendar().week).mean()['Trust_gov']

## Your Time to Shine

Come up with some new questions about this dataset and try to answer them using your newly obtained skills.